dashboard/app/models/activity.rb

```diff
@@ -72,4 +72,62 @@ def Activity.recent(limit)
     # yeah, this is a lot like .last -- but I want a dataset not an array
     Activity.order('id desc').limit(limit)
   end
+
+  # Creates a new Activity which will be written eventually to the database. (Note that the id is
+  # nil because it may not have been written yet.) An exception will be thrown if the object does
+  # not pass validation. The object is only written asynchronously if the gatekeeper allows it for
+  # this hostname.
+  def Activity.create_async!(attributes)
+    if Gatekeeper.allows('async_activity_writes', where: {hostname: Socket.gethostname})
+      activity = Activity.new(attributes)
+      activity.created_at = activity.updated_at = Time.now
+      activity.validate!
+      async_op = {'model' => 'Activity', 'action' => 'create', 'attributes' => activity.attributes}
+      activity_queue.enqueue(async_op.to_json)
+    else
+      activity = Activity.create!(attributes)
+      activity.id = nil  # Clear out the id so that no code comes to depend on it.
+    end
+    activity
+  end
+
+  # Handle an async message body created by create_async! (and other async operations we might add
+  # in the future).
+  # @param [String] json A JSON-encoded asynchronous operation.
+  def Activity.handle_async_message_json(json)
+    op = JSON::parse(json)
+
+    if op['model'] != 'Activity'
+      raise "Model must be Activity, but was #{op['model']} in #{json}"
+    end
+
+    case op['action']
+      when 'create'
+        attributes = op['attributes']
+        attributes[:updated_at] = Time.now
```

Comment:
Yes, it's needed.  The timestamps are only added when the object is saved, not when it is created.